spacy/gold.pyx

```diff
@@ -639,21 +644,246 @@ def _consume_ent(tags):
         return [start] + middle + [end]
 
 
+cdef class TokenAnnotation:
+    def __init__(self, ids=None, words=None, tags=None, heads=None, deps=None, entities=None, morphology=None, brackets=None):
+        self.ids = ids if ids else []
+        self.words = words if words else []
+        self.tags = tags if tags else []
+        self.heads = heads if heads else []
+        self.deps = deps if deps else []
+        self.entities = entities if entities else []
+        self.brackets = brackets if brackets else []
+        self.morphology = morphology if morphology else []
+
+    @classmethod
+    def from_dict(cls, token_dict):
+        return cls(ids=token_dict.get("ids", None),
+                   words=token_dict.get("words", None),
+                   tags=token_dict.get("tags", None),
+                   heads=token_dict.get("heads", None),
+                   deps=token_dict.get("deps", None),
+                   entities=token_dict.get("entities", None),
+                   morphology=token_dict.get("morphology", None),
+                   brackets=token_dict.get("brackets", None))
+
+    def to_dict(self):
+        return {"ids": self.ids,
+                "words": self.words,
+                "tags": self.tags,
+                "heads": self.heads,
+                "deps": self.deps,
+                "entities": self.entities,
+                "morphology": self.morphology,
+                "brackets": self.brackets}
+
+
+cdef class DocAnnotation:
+    def __init__(self, cats=None, links=None):
+        self.cats = cats if cats else {}
+        self.links = links if links else {}
+
+    @classmethod
+    def from_dict(cls, doc_dict):
+        return cls(cats=doc_dict.get("cats", None), links=doc_dict.get("links", None))
+
+    def to_dict(self):
+        return {"cats": self.cats, "links": self.links}
+
+
+cdef class Example:
+    def __init__(self, doc_annotation=None, token_annotations=None, doc=None,
+                 make_projective=False, ignore_misaligned=False, goldparse=None):
+        """ Doc can either be text, or an actual Doc """
+        self.doc = doc
+        self.doc_annotation = doc_annotation if doc_annotation else DocAnnotation()
+        self.token_annotations = token_annotations if token_annotations else []
```

Comment:
I think it would be better to go ahead and represent sentences differently rather than using a list of `TokenAnnotation`s here. We could have `self.token_annotation = TokenAnnotation()` to be consistent with `DocAnnotation`, a `sent_start` (or similar) attribute as a list within the `TokenAnnotation` with 0/1 or True/False values.

This would make the sentence annotation less privileged and more explicit. It's also a good bit of the work in getting a statistical sentence segmenter working, so I'd be doing it anyway in the near future, and then we don't have the awkward list here.

(I am happy to make the changes myself, but it might be easier to follow the development if this is merged into a branch here first?)

Reply:
Yea I think it's a good idea to merge this in a branch of its own so you can add your stuff on top of it. I think you have a better idea of what is needed for the statistical sentence segmenter. 

I just need to fix a few more unit tests first before we can merge this - and it'll require some more experimentation with the CLI scripts etc too, to make sure nothing broke. There's ample opportunity for bugs in this overhaul ...

Reply:
As a note to myself so I don't forget: this should actually work like `token.c.sent_start` with ternary values. It will be necessary to handle conflicting `sent_start` vs dependency parse segmentation.